### PR TITLE
Transfer function editor

### PR DESCRIPTION
**Pull request by PowerOfNames**
Adds a TransferFunction Editor GUI
Aple to:

- visually display a transferfunction and its control points on a JFreeChart inside a Swing window
- control points can be added (click), moved (drag cp) and removed (ctrl-click onto cp)
- a histogram representation of the volume data can be generated (currently 256 bins)
- histogram is dependent on the set TransferFunctionRange - also able to be edited

The GUI can be displayed via a plane (SwingUINode) using the SwingBridgeFrame (TransferFunctionEditor contains that frame)
The SwingBridgeFrame and SwingNodeUI are used to catch key inputs of scenery and trigger Swing (java.awt.KeyEvents) to simulate a click. Interaction with the 2D plane, that displayes the actual Swing window as a texture is thereby possible.

Also adds helper classes: SwingRangedSlider and SwingRangedSliderUI -> enables ranged sliders in swing (sliders with 2 buttons, that determine the value range between them)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -9,6 +9,7 @@
 .idea/
 *.iml
 *.ipr
+src/main/resources/
```
Comment (skalarproduktraum) [inline .gitignore:12]: Remove, why would you ignore the resources folder?

File: build.gradle.kts [hunk 2]
```diff
@@ -124,6 +124,11 @@ dependencies {
     testImplementation("net.imagej:imagej")
     testImplementation("net.imagej:ij")
     testImplementation("net.imglib2:imglib2-ij")
+
+    //JFreeChart
```
Comment (skalarproduktraum) [inline build.gradle.kts:128]: Remove the comments, it's clear what the packages are

File: src/main/kotlin/graphics/scenery/controls/RangeSlider.kt [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+ * Note that RangeSlider makes use of the default BoundedRangeModel, which
+ * supports an inner range defined by a value and an extent.  The upper value
+ * returned by RangeSlider is simply the lower value plus the extent.
+ */
+class RangeSlider : JSlider {
+    /**
+     * Constructs a RangeSlider with default minimum and maximum values of 0
+     * and 100.
+     */
+    constructor() {
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/controls/RangeSlider.kt:24]: Please use the Kotlin ideomatic way of declaring the default constructor.

File: src/main/kotlin/graphics/scenery/controls/RangeSliderUI.kt [hunk 4]
```diff
@@ -0,0 +1,550 @@
...
+import javax.swing.plaf.basic.BasicSliderUI
+
+/*
+From https://github.com/ernieyu/Swing-range-slider/blob/master/src/slider/RangeSliderUI.java
+ */
+/**
+ * UI delegate for the RangeSlider component.  RangeSliderUI paints two thumbs,
+ * one for the lower value and one for the upper value.
+ */
+internal class RangeSliderUI
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/controls/RangeSliderUI.kt:20]: Why is this internal?

File: src/main/kotlin/graphics/scenery/controls/RangeSliderUI.kt [hunk 5]
```diff
@@ -0,0 +1,550 @@
...
+/**
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/controls/RangeSliderUI.kt:21]: This is a strange place to add a comment, please move the Kdoc above the class definition, not between the name and the constructor brackets.

File: src/main/kotlin/graphics/scenery/controls/SwingBridgeFrame.kt [hunk 6]
```diff
@@ -0,0 +1,176 @@
+package graphics.scenery.controls
+
+
+import graphics.scenery.utils.Image
+import graphics.scenery.utils.SystemHelpers.Companion.logger
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/controls/SwingBridgeFrame.kt:5]: Wrong import, if you want to use a logger, declare a `private val logger by LazyLogger()`

File: src/main/kotlin/graphics/scenery/controls/SwingUiNode.kt [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    Vector3f(0.5f,-0.5f,0.0f),
+    Vector3f(0.5f,0.5f,0.0f)
+) {
+    var swingUiDimension = 0 to 0
+
+    init {
+        //to also hit the backside
+        this.material().cullingMode = Material.CullingMode.None
+        /*this.material {
+            textures["diffuse"] = Texture.fromImage(Image.fromStream(FileInputStream("C:\\Users\\Kodels Bier\\Desktop\\volumes\\helix.png"), ".png", false))
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/controls/SwingUiNode.kt:33]: Remove

File: src/main/kotlin/graphics/scenery/controls/SwingUiNode.kt [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    fun updateUITexture() {
+        swingBridgeFrame.finalImage?.let {
+            this.material {
+                textures["diffuse"] = Texture.fromImage((it), Texture.RepeatMode.Repeat.all(), Texture.BorderColor.OpaqueBlack, true, true,
+                    Texture.FilteringMode.NearestNeighbour, Texture.FilteringMode.NearestNeighbour, hashSetOf(Texture.UsageType.Texture))
+            }
+        }
+    }
+
+    fun Matrix4f.copy(): Matrix4f = Matrix4f(this)
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/controls/SwingUiNode.kt:53]: This is only used once and will pollute the namespace if this class is imported.

File: src/test/kotlin/graphics/scenery/tests/examples/volumes/TransferFunctionEditorExample.kt [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+ * Example for scenery - swing bridge
+ *
+ * A TransferFunctionEditor example to add, manipulate and remove control points of a volume's transfer function.
+ * Further more able to generate a histogram representation of the volume data distribution to help with the transfer function setup.
+ */
+class TransferFunctionEditorExample : SceneryBase("TransferFunctionEditor Example", 1280, 720, false) {
+    var maxCacheSize = 512
+    val cam: Camera = DetachedHeadCamera()
+
+    override fun init() {
```
Comment (skalarproduktraum) [inline src/test/kotlin/graphics/scenery/tests/examples/volumes/TransferFunctionEditorExample.kt:31]: Kdoc missing

File: src/test/kotlin/graphics/scenery/tests/examples/volumes/TransferFunctionEditorExample.kt [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+            perspectiveCamera(50.0f, windowWidth, windowHeight)
+
+            scene.addChild(this)
+        }
+
+        val workingDirectoryPath = File("").absolutePath
+
+        val options = VolumeViewerOptions().maxCacheSizeInMB(maxCacheSize)
+        val name = "RIDER"
+        val v = Volume.fromSpimData(XmlIoSpimDataMinimal().load("C:\\Users\\Kodels Bier\\Desktop\\volumes\\$name.xml"), hub, options)
```
Comment (skalarproduktraum) [inline src/test/kotlin/graphics/scenery/tests/examples/volumes/TransferFunctionEditorExample.kt:64]: Remove hardcoded path, use e.g. T1 head from the example models zipfile.

File: src/test/kotlin/graphics/scenery/tests/examples/volumes/TransferFunctionEditorExample.kt [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+            }
+        )
+        inputHandler?.addKeyBinding("ctrlClickObject", "2")
+        inputHandler?.addKeyBinding("dragObject", "1")
+    }
+
+
+    companion object {
+        @JvmStatic
+        fun main(args: Array<String>) {
```
Comment (skalarproduktraum) [inline src/test/kotlin/graphics/scenery/tests/examples/volumes/TransferFunctionEditorExample.kt:138]: Kdoc missing

File: src/test/kotlin/graphics/scenery/tests/examples/volumes/TransferFunctionEditorExample.kt [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+                    }
+                }
+            }
+        )
+        inputHandler?.addKeyBinding("ctrlClickObject", "2")
+        inputHandler?.addKeyBinding("dragObject", "1")
+    }
+
+
+    companion object {
```
Comment (skalarproduktraum) [inline src/test/kotlin/graphics/scenery/tests/examples/volumes/TransferFunctionEditorExample.kt:136]: Kdoc missing

File: src/main/kotlin/graphics/scenery/controls/SwingBridgeFrame.kt [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+            }
+            override fun keyPressed(e : KeyEvent?) {
+                updateImage()
+            }
+            override fun keyReleased(e : KeyEvent?) {
+            }
+        })
+    }
+    private fun updateImage()
+    {
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/controls/SwingBridgeFrame.kt:35]: The way this is implemented, you'll always allocate a new buffer for the image. This is probably not what you want.

File: src/main/kotlin/graphics/scenery/controls/RangeSlider.kt [hunk 14]
```diff
@@ -0,0 +1,99 @@
+package graphics.scenery.controls
+
+import javax.swing.JSlider
+
+/*
+From https://github.com/ernieyu/Swing-range-slider/blob/master/src/slider/RangeSlider.java
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/controls/RangeSlider.kt:6]: This should go into the Kdoc, together with author and license.

File: src/main/kotlin/graphics/scenery/controls/RangeSlider.kt [hunk 15]
```diff
@@ -0,0 +1,99 @@
+package graphics.scenery.controls
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/controls/RangeSlider.kt:1]: The `control` package contains the low-level stuff, like button handling, etc. I'd argue this rather belongs into a new package, e.g. named `ui` 👍

**Review by skalarproduktraum — CHANGES_REQUESTED** [review]
This will be a great addition once those few remarks are addressed 👍  Thanks, @PowerOfNames 🥳

File: src/main/kotlin/graphics/scenery/controls/SwingBridgeFrame.kt [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+ * @author Konrad Michel
+ *
+ * Bridge to link the Swing window events with scenery key inputs in combination with SwingUiNode
+ */
+open class SwingBridgeFrame(title: String) : JFrame(title) {
+
+    val uiNode = SwingUiNode(this)
+    var finalImage : Image? = null
+    var dragged = false
+    init {
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/controls/SwingBridgeFrame.kt:23]: New line missing

File: src/main/kotlin/graphics/scenery/controls/SwingBridgeFrame.kt [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+        this.addKeyListener(object : KeyListener {
+            override fun keyTyped(e : KeyEvent?) {
+            }
+            override fun keyPressed(e : KeyEvent?) {
+                updateImage()
+            }
+            override fun keyReleased(e : KeyEvent?) {
+            }
+        })
+    }
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/controls/SwingBridgeFrame.kt:33]: New line missing

File: src/test/kotlin/graphics/scenery/tests/examples/volumes/TransferFunctionEditorExample.kt (hunk 9 above)
Comment (PowerOfNames) [reply src/test/kotlin/graphics/scenery/tests/examples/volumes/TransferFunctionEditorExample.kt:31]: SHould I doc the init function? The example itself has an emplanation to it, or what am I missing here?

File: src/test/kotlin/graphics/scenery/tests/examples/volumes/TransferFunctionEditorExample.kt (hunk 11 above)
Comment (PowerOfNames) [reply src/test/kotlin/graphics/scenery/tests/examples/volumes/TransferFunctionEditorExample.kt:138]: Same question as above? None of the examples have a doc atached to the main function call

File: src/test/kotlin/graphics/scenery/tests/examples/volumes/TransferFunctionEditorExample.kt [hunk 18]
```diff
@@ -0,0 +1,161 @@
...
+
+            scene.addChild(this)
+        }
+
+        val workingDirectoryPath = File("").absolutePath
+
+        val options = VolumeViewerOptions().maxCacheSizeInMB(maxCacheSize)
+        val name = "t1-head"
+        //Currently only .xml volume formats are usable
+        val v = Volume.fromSpimData(XmlIoSpimDataMinimal().load(getDemoFilesPath() + "$name.xml"), hub, options)
```
Comment (skalarproduktraum) [inline src/test/kotlin/graphics/scenery/tests/examples/volumes/TransferFunctionEditorExample.kt:73]: This has four issues:
* the path to the file is missing a path separator
* workingDirectoryPath is unnecessary
* fromSpimData(Xml...) is also unnecessary, please use Volume.fromXML
* name could be inlined, the variable is unnecessary

File: build.gradle.kts [hunk 19]
```diff
@@ -124,6 +124,10 @@ dependencies {
     testImplementation("net.imagej:imagej")
     testImplementation("net.imagej:ij")
     testImplementation("net.imglib2:imglib2-ij")
+
```
Comment (skalarproduktraum) [inline build.gradle.kts:127]: One newline too many

File: src/main/kotlin/graphics/scenery/UI/RangeSlider.kt [hunk 20]
```diff
@@ -0,0 +1,88 @@
+package graphics.scenery.UI
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/UI/RangeSlider.kt:1]: Package names are always lowercase (by convention)

File: src/main/kotlin/graphics/scenery/UI/SwingBridgeFrame.kt [hunk 21]
```diff
@@ -0,0 +1,201 @@
+package graphics.scenery.UI
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/UI/SwingBridgeFrame.kt:1]: Package names are always lowercase (by convention)

File: src/main/kotlin/graphics/scenery/UI/SwingUiNode.kt [hunk 22]
```diff
@@ -0,0 +1,96 @@
+package graphics.scenery.UI
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/UI/SwingUiNode.kt:1]: Package names are always lowercase (by convention)

File: src/main/kotlin/graphics/scenery/UI/SwingUiNode.kt [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+import graphics.scenery.attribute.material.Material
+import graphics.scenery.primitives.Plane
+import graphics.scenery.textures.Texture
+import org.joml.Matrix4f
+import org.joml.Vector2f
+import org.joml.Vector3f
+import org.joml.Vector4f
+
+/**
+ * @author Jan Tiemann
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/UI/SwingUiNode.kt:12]: Author names should have an email address as well

**Review by skalarproduktraum — CHANGES_REQUESTED** [review]
A little bit more nitpicking, sorry ^^

**Comment by skalarproduktraum** [comment]
Also, please check why Gitlab is failing 👍

File: build.gradle.kts (hunk 19 above)
Comment (PowerOfNames) [reply build.gradle.kts:127]: xD

File: src/main/kotlin/graphics/scenery/ui/SwingBridgeFrame.kt [hunk 24]
```diff
@@ -0,0 +1,191 @@
...
+                }
+            }
+
+            override fun keyTyped(e : KeyEvent?) {}
+            override fun keyReleased(e : KeyEvent?) {}
+        })
+    }
+
+    /**
+     * Creates a snapshot of the actual SwingFrame and converts it to a ByteBuffer. Tehn updates the texture of the 2DPlane that renders in the scene to present
```
Comment (moreApi) [inline src/main/kotlin/graphics/scenery/ui/SwingBridgeFrame.kt:56]: then* typo

File: src/main/kotlin/graphics/scenery/ui/SwingBridgeFrame.kt [hunk 25]
```diff
@@ -0,0 +1,191 @@
...
+ * Bridge to link the Swing window events with scenery key inputs in combination with SwingUiNode
+ */
+open class SwingBridgeFrame(title: String) : JFrame(title) {
+
+    val uiNode = SwingUiNode(this)
+
+    private var dragged = false
+    private var lastUpdate = 0L
+    private var snapshotBuffer : ByteBuffer? = null
+    var finalImage : Image? = null
```
Comment (moreApi) [inline src/main/kotlin/graphics/scenery/ui/SwingBridgeFrame.kt:29]: this is not a fitting name for this as a property. Maybe currentImage, currentSnapshot or just snapshot are more fitting
the terminology should be the same with the "snapshotBuffer"

File: src/main/kotlin/graphics/scenery/volumes/HasHistogram.kt [hunk 26]
```diff
@@ -0,0 +1,8 @@
+package graphics.scenery.volumes
+
+import net.imglib2.histogram.Histogram1d
+
+interface HasHistogram {
```
Comment (moreApi) [inline src/main/kotlin/graphics/scenery/volumes/HasHistogram.kt:5]: li'l docu pls

File: src/main/kotlin/graphics/scenery/volumes/HasTransferFunction.kt [hunk 27]
```diff
@@ -0,0 +1,8 @@
+package graphics.scenery.volumes
+
+interface HasTransferFunction {
```
Comment (moreApi) [inline src/main/kotlin/graphics/scenery/volumes/HasTransferFunction.kt:3]: li'l docu pls

**Review by moreApi — CHANGES_REQUESTED** [review]

**Review by moreApi — APPROVED** [review]

File: src/test/kotlin/graphics/scenery/tests/examples/volumes/TransferFunctionEditorExample.kt [hunk 28]
```diff
@@ -0,0 +1,156 @@
...
+ * @author Konrad Michel <Konrad.Michel@mailbox.tu-dresden.de>
+ * Example for scenery - swing bridge
+ *
+ * A TransferFunctionEditor example to add, manipulate and remove control points of a volume's transfer function.
+ * Further more able to generate a histogram representation of the volume data distribution to help with the transfer function setup.
+ *
+ * Usage: To enable the UI on the plane click once (Key '1') while hovering over the plane. Key '1' used as normal Mouse-interactions (Clicking and dragging).
+ * Control Points can be dragged, added and removed. A remove happens via Ctrl-Clicking (In this example managed by using Key '2'.
+ */
+class TransferFunctionEditorExample : SceneryBase("TransferFunctionEditor Example", 1280, 720, false) {
```
Comment (moreApi) [inline src/test/kotlin/graphics/scenery/tests/examples/volumes/TransferFunctionEditorExample.kt:28]: please put "@JvmOverloads" in line 27 so it is more convinient to use in Java

**Review by moreApi — CHANGES_REQUESTED** [review]

File: src/main/kotlin/graphics/scenery/ui/SwingUiNode.kt [hunk 29]
```diff
@@ -0,0 +1,96 @@
...
+import org.joml.Vector3f
+import org.joml.Vector4f
+
+/**
+ * @author Jan Tiemann <j.tiemann@hzdr.de>
+ * @author Konrad Michel <Konrad.Michel@mailbox.tu-dresden.de>
+ *
+ * Scene representation of a SwingFrame on a plane in form of a texture snapshot render of the frame in combination with SwingBridgeFrame
+ */
+class SwingUiNode(private val swingBridgeFrame : SwingBridgeFrame) : Plane(
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/ui/SwingUiNode.kt:17]: A minor point, but in terms of consistency, `SwingUINode` would make more sense 👍

**Review by skalarproduktraum — CHANGES_REQUESTED** [review]
Tiny nitpicking, looks great otherwise. Big thanks @PowerOfNames and @moreApi, this is extremely useful functionality!

File: src/main/kotlin/graphics/scenery/ui/SwingUiNode.kt [hunk 30]
```diff
@@ -0,0 +1,96 @@
...
+        }
+    }
+
+    /**
+     * Function that takes the last taken screenshot from the SwingUI-window coming from the attached swingBridgeFrame and updates the SwingUINodes-Plane texture with it
+     */
+    fun updateUITexture() {
+        swingBridgeFrame.finalImage?.let {
+            this.material {
+                textures["diffuse"] = Texture.fromImage((it), Texture.RepeatMode.Repeat.all(), Texture.BorderColor.OpaqueWhite, true, true,
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/ui/SwingUiNode.kt:44]: There's no `it` here, this is why the CI fails currently. Please fix 👍

**Review by skalarproduktraum — CHANGES_REQUESTED** [review]
And another mistake identified that makes the CI fail.

**Review by skalarproduktraum — APPROVED** [review]

**Pull request merged**